# Supplementary Tables
* Table S1. Genomes used for ISCompare evaluation
* Table S2. SurroundingLen parameter optimization. Sheet 1, results of the comparison of E. coli K-12 substr. MG1655 with an artificial genome of the same strain containing 100 IS30 random insertions. Sheet 2, results of the comparison using E. meliloti strain 1021 as reference genome and an artificial E. meliloti 2011 genome with 100 ISRm5 random insertions as query. Sheet 3, Sensitivity and precision analysis. TP, true positives, FP, false positives, TP*, true positives manually found by inspecting the VD reports. 
* Table S3. ISCompare evaluation using 3,000 random IS30 insertions. Sheet 1: Statistical analysis. Sheet 2: Location of 3,000 randomly inserted IS30. Set 1. Sheet 3: ISCompare results using a compilation of IS. Sheet 4: Location of 3,000 randomly inserted IS30. Set 2. Sheet 5: ISCompare results using IS30 as query IS. TP, true positives, FP, false positives, TP*, true positives manually found by inspecting the VD reports. 
* Table S4. Analysis of differentially located ISs on P. aeruginosa strains. Sheet 1: ISCompare result analysis. TP, true positives, FP, false positives, TP*, true positives manually found by inspecting the VD reports. Sheet 2: ANIb and DDH results.
* Table S5. Analysis of differentially located ISs on E. meliloti strains. Sheet 1: ISCompare result analysis. TP, true positives, FP, false positives, TP*, true positives manually found by inspecting the VD reports. Sheet 2: ANIb calculated using ANI matrix calculator server. Sheet 3: dDDH results from http://ggdc.dsmz.de/ggdc.php.
* Table S6. Comparison of ISCompare results using the normal vs the Shift mode (-S). A comparison of DLIS in E. meliloti 1021 and GR4 strains was done using ISCompare with -S option set to 5,000 nucleotides. 
* Table S7. Comparison of ISCompare and ISSeeker using E. meliloti genomes. E. meliloti 1021 was used as reference, and compared to GR4 and U1022 strains as queries. In the case of GR4 strain, the -S option was also evaluated. As ISSeeker only can analyse one IS at a time, only ISRm2011-2 location was analysed using both programs. Sheet 1: ISCompare vs ISSeeker results summary. Sheet 2: ISSeeker, 1021 vs GR4 results. Sheet 3: ISCompare,  1021 vs GR4 results. Sheet 4: ISCompare with -S 5000 setting. 1021 vs GR4 results. Sheet 5: ISSeeker, 1021 vs USDA1022 results. Sheet 6: ISCompare. 1021 vs USDA1022 results. Sheet 7: ISCompare with -S 5000 setting. 1021 vs USDA1022 results.
* Table S8. Comparison of ISCompare and ISSeeker using P. aeruginosa genomes. Sheet 1: ISCompare vs ISSeeker results summary. Sheet 2: ISSeeker results for all the analysed P. aeruginosa strains. Sheet 3:  results for all the analysed P. aeruginosa strains.
* Table S9. Comparison of B. pertussis TOHAMA I with strains BpH915 and BpI127 containing a IS481 insertion on the pertactin autotransporter gene. Sheet 1: Results of TOHAMA I vs BpH915 using ISCompare. Sheet 2: Results of TOHAMA I vs BpI127 using ISCompare.
